### DISCUSSION TITLE
Should external authentication scripts defend against timing attacks to detect if a user exists?

### DISCUSSION THREAD

**Discussion by dseomn** [processone/ejabberd#3679] [Questions]
https://docs.ejabberd.im/developer/guide/#external has both `auth` to test if a password is valid for a user, and `isuser` to test if a user exists. Normally when testing if a password is valid for a user, it's a good idea to prevent a [timing attack](https://en.wikipedia.org/wiki/Timing_attack) by taking the same amount of time whether the user exists or not. But depending on how ejabberd calls `isuser`, it might not be worth avoiding that timing attack. Does anybody know if it's worth defending against that in `auth`, or if `isuser` provides a much bigger side channel?

**Comment by licaon-kter** [comment]
You are attacking yourself? ejabberd calls the script on user login, maybe you can throttle that before auth: https://docs.ejabberd.im/admin/configuration/basic/#shaper-rules

**Reply by dseomn** [reply]
> You are attacking yourself?

No, I'm trying to figure out what an external authentication script needs to do to prevent attacks by anybody with access to the client-to-server port. My impression is that those ports are often exposed to the public internet.

> ejabberd calls the script on user login, maybe you can throttle that before auth: https://docs.ejabberd.im/admin/configuration/basic/#shaper-rules

I don't think that would help by itself, since the timing difference between testing if a user exists and checking the user's password(s) is probably significant enough to attack with only a handful of attempts. It also requires additional configuration by anybody who uses the external authentication script in their own server.

**Comment by prefiks** [comment]
Please don't add delay for isuser, this can be triggered quite a bit in code that has nothing to do with user registration.

**Reply by dseomn** [reply]
I'm asking about adding delay in auth, not isuser.
